R/backend-snowflake.R

```diff
@@ -18,13 +18,193 @@ NULL
+      # so we must set <occurrence> to 1 if not replacing all.  See:
+      # https://docs.snowflake.com/en/sql-reference/functions/regexp_replace.html
+      # Also, Snowflake needs backslashes escaped, so we must increase the
+      # level of escaping by 1
+      str_replace = function(string, pattern, replacement) {
+        pattern <- gsub("\\\\", "\\\\\\\\", pattern)
+        sql_expr(regexp_replace(!!string, !!pattern, !!replacement, 1, 1))
+      },
+      str_replace_all = function(string, pattern, replacement) {
+        pattern <- gsub("\\\\", "\\\\\\\\", pattern)
+        sql_expr(regexp_replace(!!string, !!pattern, !!replacement))
+      },
+      str_remove = function(string, pattern) {
+        pattern <- gsub("\\\\", "\\\\\\\\", pattern)
+        sql_expr(regexp_replace(!!string, !!pattern, "", 1, 1))
+      },
+      str_remove_all = function(string, pattern) {
+        pattern <- gsub("\\\\", "\\\\\\\\", pattern)
+        sql_expr(regexp_replace(!!string, !!pattern))
+      },
+      str_trim = function(string) {
+        sql_expr(trim(!!string))
+      },
+      str_squish = function(string) {
+        sql_expr(regexp_replace(trim(!!string), "\\\\s+", " "))
+      },
+
+
+      # lubridate functions
+      # https://docs.snowflake.com/en/sql-reference/functions-date-time.html
+      day = function(x) {
+        sql_expr(EXTRACT(DAY %FROM% !!x))
+      },
+      mday = function(x) {
+        sql_expr(EXTRACT(DAY %FROM% !!x))
+      },
+      wday = function(x, label = FALSE, abbr = TRUE, week_start = NULL) {
+        if (!label) {
+          week_start <- week_start %||% getOption("lubridate.week.start", 7)
+          offset <- as.integer(7 - week_start)
+          sql_expr(EXTRACT("dayofweek", DATE(!!x) + !!offset) + 1)
+        } else if (label && !abbr) {
+          sql_expr(
+            DECODE(
+              EXTRACT("dayofweek", !!x),
+              1, "Monday",
+              2, "Tuesday",
+              3, "Wednesday",
+              4, "Thursday",
+              5, "Friday",
+              6, "Saturday",
+              0, "Sunday"
+            )
+          )
+        } else if (label && abbr) {
+          sql_expr(DAYNAME(!!x))
+        } else {
+          abort("Unrecognized arguments to `wday`")
+        }
+      },
+      yday = function(x) sql_expr(EXTRACT("dayofyear", !!x)),
+      week = function(x) {
+        sql_expr(FLOOR((EXTRACT("dayofyear", !!x) - 1L) / 7L) + 1L)
+      },
+      isoweek = function(x) sql_expr(EXTRACT("weekiso", !!x)),
+      # For consistency with postgres, right-pad month names for uniform size
+      month = function(x, label = FALSE, abbr = TRUE) {
+        if (!label) {
+          sql_expr(EXTRACT("month", !!x))
+        } else {
+          if (abbr) {
+            sql_expr(MONTHNAME(!!x))
+          } else {
+            sql_expr(
+              DECODE(
+                EXTRACT("month", !!x),
+                1, "January  ",
```

Comment:
Should these have trailing spaces?

Reply:
I was trying to be consistent with the postgres implementation, which uses `TO_CHAR(., "Month")`, which yields
> full upper case month name (blank-padded to 9 chars)

As described here: https://www.postgresql.org/docs/current/functions-formatting.html

But maybe this isn't a good reason; I'll strip the whitespace in my update.